src/daemon/daemon.cpp

```diff
@@ -801,6 +804,70 @@ std::string generate_unused_mac_address(std::unordered_set<std::string>& s)
                     max_tries, s.size())};
 }
 
+std::string run_in_vm(mp::SSHSession& session, const std::string& cmd)
+{
+    auto proc = session.exec(cmd);
+    if (proc.exit_code() != 0)
+    {
+        auto error_msg = proc.read_std_error();
+        mpl::log(mpl::Level::warning, category,
+                 fmt::format("failed to run '{}', error message: '{}'", cmd, mp::utils::trim_end(error_msg)));
+        return std::string{};
+    }
+
+    auto output = proc.read_std_output();
+    if (output.empty())
+    {
+        mpl::log(mpl::Level::warning, category, fmt::format("no output after running '{}'", cmd));
+        return std::string{};
+    }
+
+    return mp::utils::trim_end(output);
+}
+
+bool is_ipv4_valid(const std::string& ipv4)
+{
+    try
+    {
+        (mp::IPAddress(ipv4));
+    }
+    catch (std::invalid_argument&)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+std::vector<std::string> get_all_ipv4(mp::SSHSession& session)
+{
+    std::vector<std::string> all_ipv4;
+
+    auto ip_a_output = QString::fromStdString(run_in_vm(session, "ip a"));
```

Comment:
I was in doubt about the link scope. I decided to include the addresses with link scope because I don't see why a user can't configure in the instance a private address reachable only from the host. Am I wrong?

Reply:
Anyway, I left only global scope addresses. If needed, it'll be easy to add the link scope addresses.

Reply:
I am not entirely sure either but, according to [documentation](http://linux-ip.net/html/tools-ip-address.html), link scope here means "valid only on this device". I think it is only useful for special addresses (e.g. broadcast) or perhaps other special routing stuff.

Reply:
Ok, let's leave only global. scope, then.